srml/support/src/storage/generator/linked_map.rs

```diff
@@ -0,0 +1,289 @@
+		Linkage::default()
+	}
+}
+
+/// Read current head pointer.
+fn read_head<K, V, G>() -> Option<K>
+where
+	K: Codec,
+	V: Codec,
+	G: StorageLinkedMap<K, V>
+{
+	unhashed::get(G::final_head_key())
+}
+
+/// Overwrite current head pointer.
+///
+/// If `None` is given head is removed from storage.
+fn write_head<K, V, G>(head: Option<&K>)
+where
+	K: Codec,
+	V: Codec,
+	G: StorageLinkedMap<K, V>
+{
+	match head {
+		Some(head) => unhashed::put(G::final_head_key(), head),
+		None => unhashed::kill(G::final_head_key()),
+	}
+}
+
+impl<K: Codec, V: Codec, G: StorageLinkedMap<K, V>> storage::StorageLinkedMap<K, V> for G {
+	type Query = G::Query;
+
+	fn exists<KeyArg: Borrow<K>>(key: KeyArg) -> bool {
+		unhashed::exists(Self::storage_linked_map_final_key(key).as_ref())
+	}
+
+	fn get<KeyArg: Borrow<K>>(key: KeyArg) -> Self::Query {
+		let val = unhashed::get(Self::storage_linked_map_final_key(key).as_ref());
+		G::from_optional_value_to_query(val)
+	}
+
+	fn insert<KeyArg: Borrow<K>, ValArg: Borrow<V>>(key: KeyArg, val: ValArg) {
+		let final_key = Self::storage_linked_map_final_key(key.borrow());
+		let linkage = match read_with_linkage::<_, _, G>(final_key.as_ref()) {
+			// overwrite but reuse existing linkage
+			Some((_data, linkage)) => linkage,
+			// create new linkage
+			None => new_head_linkage::<_, _, G>(key.borrow()),
+		};
+		unhashed::put(final_key.as_ref(), &(val.borrow(), linkage))
+	}
+
+	fn insert_ref<KeyArg: Borrow<K>, ValArg: ?Sized + Encode>(key: KeyArg, val: &ValArg)
+	where
+		V: AsRef<ValArg>
+	{
+		let final_key = Self::storage_linked_map_final_key(key.borrow());
+		let linkage = match read_with_linkage::<_, _, G>(final_key.as_ref()) {
+			// overwrite but reuse existing linkage
+			Some((_data, linkage)) => linkage,
+			// create new linkage
+			None => new_head_linkage::<_, _, G>(key.borrow()),
+		};
+		unhashed::put(final_key.as_ref(), &(&val, &linkage))
+	}
+
+	fn remove<KeyArg: Borrow<K>>(key: KeyArg) {
+		G::take(key);
+	}
+
+	fn mutate<KeyArg: Borrow<K>, R, F: FnOnce(&mut Self::Query) -> R>(key: KeyArg, f: F) -> R {
+		let final_key = Self::storage_linked_map_final_key(key.borrow());
+
+		let (mut val, _linkage) = read_with_linkage::<_, _, G>(final_key.as_ref())
+			.map(|(data, linkage)| (G::from_optional_value_to_query(Some(data)), Some(linkage)))
+			.unwrap_or_else(|| (G::from_optional_value_to_query(None), None));
+
+		let ret = f(&mut val);
+		match G::from_query_to_optional_value(val) {
+			Some(ref val) => G::insert(key.borrow(), val),
+			None => G::remove(key.borrow()),
+		}
+		ret
+	}
+
+	fn take<KeyArg: Borrow<K>>(key: KeyArg) -> Self::Query {
+		let final_key = Self::storage_linked_map_final_key(key);
+
+		let full_value: Option<(V, Linkage<K>)> = unhashed::take(final_key.as_ref());
+
+		let value = full_value.map(|(data, linkage)| {
+			remove_linkage::<_, _, G>(linkage);
+			data
+		});
+
+		G::from_optional_value_to_query(value)
+	}
+
+	fn enumerate() -> Box<dyn Iterator<Item = (K, V)>>
```

Comment:
Can we not return `Enumerator` directly here? Without boxing?

Reply:
yeah but for that we need to make Enumerator as a generic parameter of the trait StorageLinkedMap<K, V, Enumerator>

because we can't have `fn enumerate() -> impl Iterator<Item = (K, V)> where K: 'static, V: 'static, Self: 'static;` in the trait definition

This can be done indeed.

Reply:
I mean I don't mind either implementation, if you say you prefer one I go for it.

Reply:
YEah, we could just add it as associated type or as generic parameter. I would like to prevent to allocate :)

Reply:
done :-)